rllib/algorithms/dqn/torch/dqn_rainbow_torch_rl_module.py

```diff
@@ -0,0 +1,319 @@
+from typing import Dict, List, Tuple, Union
+
+from ray.rllib.algorithms.dqn.dqn_rainbow_rl_module import (
+    DQNRainbowRLModule,
+    ATOMS,
+    QF_LOGITS,
+    QF_PROBS,
+    QF_TARGET_NEXT_PREDS,
+    QF_TARGET_NEXT_PROBS,
+)
+from ray.rllib.algorithms.sac.sac_rl_module import QF_PREDS
+from ray.rllib.core.models.base import Encoder, ENCODER_OUT, Model
+from ray.rllib.core.rl_module.torch.torch_rl_module import TorchRLModule
+from ray.rllib.core.rl_module.rl_module import RLModule
+from ray.rllib.core.rl_module.rl_module_with_target_networks_interface import (
+    RLModuleWithTargetNetworksInterface,
+)
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.typing import NetworkType, TensorType, TensorStructType
+
+torch, nn = try_import_torch()
+
+
+class DQNRainbowTorchRLModule(TorchRLModule, DQNRainbowRLModule):
+    framework: str = "torch"
+
+    @override(RLModule)
+    def _forward_inference(self, batch: Dict[str, TensorType]) -> Dict[str, TensorType]:
+        output = {}
+
+        # Encoder forward pass.
+        encoder_outs = self.encoder(batch)
+
+        # Q-head.
+        qf_outs = self._qf(encoder_outs[ENCODER_OUT])
+
+        # Get action distribution.
+        action_dist_cls = self.get_exploration_action_dist_cls()
+        action_dist = action_dist_cls.from_logits(qf_outs[QF_PREDS])
+        # Note, the deterministic version of the categorical distribution
+        # outputs directly the `argmax` of the logits.
+        exploit_actions = action_dist.to_deterministic().sample()
+
+        # In inference we only need the exploitation actions.
+        output[SampleBatch.ACTIONS] = exploit_actions
+
+        return output
+
+    # TODO (simon): Implement the timestep for more complex exploration
+    # schedules with epsilon greedy.
+    @override(RLModule)
+    def _forward_exploration(
+        self, batch: Dict[str, TensorType]
+    ) -> Dict[str, TensorType]:
+        output = {}
+
+        # Q-network forward pass.
+        qf_outs = self._qf(batch)
+
+        # Get action distribution.
+        action_dist_cls = self.get_exploration_action_dist_cls()
+        action_dist = action_dist_cls.from_logits(qf_outs[QF_PREDS])
+        # Note, the deterministic version of the categorical distribution
+        # outputs directly the `argmax` of the logits.
+        exploit_actions = action_dist.to_deterministic().sample()
+
+        # Apply epsilon-greedy exploration.
+        # TODO (simon): Implement sampling for nested spaces.
```

Comment:
@sven1977 How is this actually best managed in the new stack? Do we support this already?

Reply:
We don't yet. After SAC/DQN/APEX-DQN are done, we should write an off-policy algo that can handle nested and arbitrary action spaces.

Reply:
Didn't we have also online algorithms with complex action spaces in the old stack?